feat: Integrate new Collage Designer and replace old Generator

dasGOLE opened the pull request:
This commit introduces a new, dedicated Collage Designer, replacing the legacy collage generator. The goal is to provide a more flexible and robust platform for designing collages and also be able to move all collage settings in this designer in the future, laying the groundwork for future enhancements and a better user experience.

<!--
    Thank you for contributing!
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/PhotoboothProject/photobooth/blob/dev/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "x" next to an item)
<!-- Example:
- [x] Bug fix
-->

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Other, please explain:

<!--
    Please ensure your pull request is ready:

    - Please run 'npm run build' before submitting to have consistent formatting for both JavaScript & SCSS
    - Please run 'npm run eslint' to make sure there's no lint issues
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)
Key changes include:

- **Removal of Legacy Collage Generator:** The old generator's entry points and related files have been removed or redirected. (Note: A dedicated cleanup commit for all generator-related files is planned for the future.)
- **New Collage Designer Structure:** A dedicated directory and file structure for the new Collage Designer (`admin/collage-designer/`) have been established to encapsulate its logic and assets.
- **Initial Collage Designer Layout:** The foundational HTML/CSS layout for the Collage Designer has been created, providing a basic UI for future design capabilities.
- **Unified Configuration Rendering:** The `renderConfigManager` was introduced and integrated into `inputAdmin.php` to standardize the rendering of configuration options across the admin panel, improving consistency and maintainability.
- **Theme Renderer Adaptation:** The theme renderer has been updated to utilize the new, general `renderConfigManager`, ensuring that theme-specific configurations are also processed through the unified rendering system.
- **Enhanced Admin Panel Navigation:**
    - The "Collage Designer" button in the admin panel now directly navigates to the new designer within the same tab, providing a smoother user flow.
- **Optimized Admin Settings Save Logic:**
    - A centralized `saveAdminSettings()` JavaScript function has been implemented for handling API requests to save admin configurations.
    - This function uses Promises for asynchronous handling and allows configurable behavior (e.g., page reload) via options.
    - An `isDirty` check (based on `$('#save-admin-btn').hasClass('isDirty')`) ensures that settings are only saved if actual changes have occurred, preventing unnecessary API calls and improving responsiveness.
    - The "Collage Designer" button now conditionally saves admin settings before navigation *only if* there are pending changes, otherwise, it navigates directly.
    - The main "Save" button in the admin panel retains its original behavior of reloading the page after saving.

#### Is there anything you'd like reviewers to focus on?

dasGOLE committed:
feat: Integrate new Collage Designer and replace old Generator

This commit introduces a new, dedicated Collage Designer, replacing the legacy collage generator. The goal is to provide a more flexible and robust platform for designing collages and also be able to move all collage settings in this designer in the future, laying the groundwork for future enhancements and a better user experience.

Key changes include:

- **Removal of Legacy Collage Generator:** The old generator's entry points and related files have been removed or redirected. (Note: A dedicated cleanup commit for all generator-related files is planned for the future.)
- **New Collage Designer Structure:** A dedicated directory and file structure for the new Collage Designer (`admin/collage-designer/`) have been established to encapsulate its logic and assets.
- **Initial Collage Designer Layout:** The foundational HTML/CSS layout for the Collage Designer has been created, providing a basic UI for future design capabilities.
- **Unified Configuration Rendering:** The `renderConfigManager` was introduced and integrated into `inputAdmin.php` to standardize the rendering of configuration options across the admin panel, improving consistency and maintainability.
- **Theme Renderer Adaptation:** The theme renderer has been updated to utilize the new, general `renderConfigManager`, ensuring that theme-specific configurations are also processed through the unified rendering system.
- **Enhanced Admin Panel Navigation:**
    - The "Collage Designer" button in the admin panel now directly navigates to the new designer within the same tab, providing a smoother user flow.
- **Optimized Admin Settings Save Logic:**
    - A centralized `saveAdminSettings()` JavaScript function has been implemented for handling API requests to save admin configurations.
    - This function uses Promises for asynchronous handling and allows configurable behavior (e.g., page reload) via options.
    - An `isDirty` check (based on `$('#save-admin-btn').hasClass('isDirty')`) ensures that settings are only saved if actual changes have occurred, preventing unnecessary API calls and improving responsiveness.
    - The "Collage Designer" button now conditionally saves admin settings before navigation *only if* there are pending changes, otherwise, it navigates directly.
    - The main "Save" button in the admin panel retains its original behavior of reloading the page after saving.

Change-Id: Iecdac17ee959247347cfef99f687e1e93431d55d

dasGOLE committed:
feat: Introduce dedicated Back button component and admin input renderer

New back button functionality in a dedicated helper file (`admin/helper/backBtn.php`). This new component provides a robust "Back" button experience by leveraging the browser's `history.back()` JavaScript API, ensuring navigation to preceding page in the browser history, including hash fragments.

Additionally, an `AdminInput` renderer was created.

Key changes include:
- **New File:** `admin/helper/backBtn.php` containing the `getBackBtn` function.
- ** Back Button Logic:** The `getBackBtn` function uses `href="javascript:history.back()"` and `onclick="event.preventDefault(); history.back();"` for reliable client-side navigation.
- **Admin Input Renderer:** Implemented a new renderer for the back button, allowing it to be easily added to admin forms and layouts.

Change-Id: I79e2d12f3c5f6e4dfebf5c1025f74a2fc8b48c69

dasGOLE committed:
fixed formatting and removed "from" tag from collage-desginer Button

Change-Id: I487e4a1a755d2f2a10ef1f47d7e1aa2d1cef3772

dasGOLE committed:
feat: Implement grouped collage layout selection

Key changes include:
- **new CollageLayoutScanner:** new scanner to dynamically scan all available json-designs.
- **Nested Grouping:** Collage layouts grouped into two levels (e.g., "Standard Layouts" > "Portrait Layouts", "Custom Layouts" > "Community Layouts") for better organization and clarity.
- **Dynamic Directory Creation:** The scanner automatically creates missing base (`private/collage`) and subdirectories (`private/collage/portrait`, `private/collage/community`) if they don't exist.
- **Robust Layout Naming:** Each layout is guaranteed to have a display name, using the `name` field from its JSON configuration or falling back to the filename (`layoutId`) if `name` is missing.
- **changes Collage template:** added `name` field

dasGOLE committed:
collage-desinger: small fixes

dasGOLE committed:
feat(collage): Adapt layout selection and limit calculation for new designer

Key changes include:
- **Adapted Layout Selection:** The collage layout selection mechanism has been updated to align with the new designer's data structure. This means `config['collage']['layout']` now stores the relative path to the layout's JSON configuration file instead of an enum value.
- **Persistent Collage Configuration:** The calculation of the collage limit and other related settings (e.g., placeholder properties) are now always saved to the main configuration, regardless of whether the collage feature is currently enabled. This ensures that the configuration is consistently up-to-date and prevents issues where stale limit values were used, leading to "Undefined array key" errors.
- **Removed `CollageLayoutEnum`:** The `CollageLayoutEnum` and all its references have been deprecated and removed, simplifying the code and eliminating unnecessary abstraction now that layout paths are directly used.
- **Enhanced `AdminInput::renderSelect`:** The `renderSelect` method in `AdminInput` now supports providing an HTML string as an option list, allowing for more flexible and dynamic dropdown constructions (e.g., with `optgroup` elements).
- **Extended `CollageLayoutScanner`:** The `CollageLayoutScanner` class has been extended with new utility methods:
    - `getLayoutData()`: To retrieve detailed data for a specific layout.
    - `getLayoutSelectOptionsHtml()`: To generate the formatted HTML options for the select dropdown, including `optgroup` for better organization.
    - `getCollageConfigPath()`: To correctly determine the full file path for a given collage layout.

reloxx13 committed:
fix rebase missin brace, fix cs and ci

reloxx13 commented:
rebased and cs/ci fixes

reloxx13 committed:
fix ci

dasGOLE committed:
feat: collage-designer:
**draw current layout to canvas
**drag and drop placeholder (translation)

dasGOLE committed:
feat: collage-designer:
- add handles for interactive scaling

dasGOLE committed:
feat: collage-designer
- add handle for interactive rotation

dasGOLE committed:
feat: collage-designer: added rotation icon from `feathericons.com´

dasGOLE committed:
collage-designer: added alaignment tools

dasGOLE committed:
collage-desginer: implementation of alaignment tools

dasGOLE committed:
collage-designer: added grouped moving, rotation and scaling.
- removed unproportional scaling (didn't work reliable)

dasGOLE committed:
collage-designer: added functions for distribution buttons

dasGOLE committed:
collage-designer: added Undo/Redo Buttons and functionality

dasGOLE committed:
collage-designer: added Undo/Redo with Keyboard Shortcuts

dasGOLE committed:
collage-designer: added `add button` for new placeholders and functionality

dasGOLE committed:
collage-designer: dynamic loading of demo Img

dasGOLE committed:
collage-designer: added functionality for deleting objekts

dasGOLE committed:
collage-designer: made Handels responsive and added delete icon for mouse hover

dasGOLE committed:
collage-designer: added z-order functions

dasGOLE committed:
collage-designer:
- refractored some of the Designer-Layout
- added new .js for the element-settings-panel (`collage-designer-elemntSetPnl.js`)
- rewritten `element-settings-panel.php`
- changed dragging behavior, to ensure no dragging while selecting

dasGOLE committed:
collage-designer: element-settings-panel: Limit input fields

dasGOLE committed:
collage-designer: removed backspace for removing elements

dasGOLE committed:
collage-designer: changed rotation via mouse to 1° increment

dasGOLE committed:
collage-designer: added `lock aspect ratio` funktionality and keep aspect ratio while scaling

Still not working: when pressing `shift` while scaling, snap back to original aspect ratio

dasGOLE committed:
collage-designer: refactored img and text specific setting panel.

Added first functions for img settings panel

reloxx13 committed:
fix dont save admin if no changes

reloxx13 committed:
order en translation

reloxx13 committed:
add gtRelativePath and deprecate bad function name

reloxx13 commented:
I did my best to rebase and merge it with the new collage selector, please check if everything is working as expected.
I also wrote a migration from old selector to new, please also check it.

TODO:
The new collage selector lacks support for multiple custom collage support from the private/collage dir.
For this I prepared but out commented a migration.
The ENUM makes Problems here `Uncaught ValueError: "private/collage/community/collage.json" is not a valid backing value for enum Photobooth\Enum\CollageLayoutEnum`. Dunno how to handle it right now cause we can habe multiple custom collage jsons now in diffrent directories.
This is in conflict because the predefined collages are a enum now

reloxx13 committed:
fix cs ci

reloxx13 committed:
add german translation

reloxx13 committed:
fix button translation

reloxx13 committed:
translation, fix open designer button

dasGOLE commented:
Hey,
thanks a lot for your work on this.
At the moment, I still have quite a few items on my TODO list for the designer – including, towards the end, all translations.
Since quite a lot has changed on the dev side regarding collages in the meantime, I deliberately didn’t plan a rebase for now. My idea was to wait until the planned features are fully integrated and then think about how the designer should be merged with the backend / the rest of the system. Because of this, I also removed the enum, as it prevents having a dynamic selection for new designs.
In general, I’m in favor of keeping the back button and not opening the designer in a new tab. This would also allow us to finally clean up the admin panel by introducing subpages for settings. In the long run, all collage-related settings should live in the designer, with only a few quick settings remaining in the admin panel if necessary.

dasGOLE committed:
colllage-designer:
- added aspect ratio functionality
- added the property 'type' to canvas elements
- some refactoring

dasGOLE committed:
collage-designer: changed the icon for the rotating hover effect

dasGOLE committed:
collage-designer: Changes to apply aspect ratio

dasGOLE committed:
collage-designer: some refactoring

dasGOLE committed:
collage-designer: added frame functions for single image in collage

dasGOLE committed:
collage-designer:: added placeholder functionallity

new behavior: placeholder are now active, if a path is present in `elements.src` in the JSON file. Therefore more than one placeholder can be configured and even diffrent images!
(needs additional changes in the backend!)

dasGOLE committed:
collage-designer: fix undo / redo for placeholder

- changed icon for placeholder in img-settings-panel
- changed padding for `remove placeholder`button

dasGOLE committed:
collage-designer: added text fields

- added button to add text fields
- added buttons, functions and element attributes to enable bold, italic and underlined text
- added buttons and functions for applying alignment to the text
- some refactoring

dasGOLE committed:
collage-designer: added dynamic found loading

dasGOLE committed:
collage-designer:  fix `selected element` text in element settings panel

dasGOLE committed:
collage-designer:  added functions to edit text, textcolor and font

dasGOLE committed:
collage-designer:: added button to toggle the outline of all elements

dasGOLE committed:
collage-designer: small changes

dasGOLE committed:
collage-designer: The DrawCanvas and MouseHandles functions have been rewritten. A new drawHandles function has been added to centralise all Handle actions.

dasGOLE committed:
collage-designer: preview_titel shows selected Layout name.

- Back button is now aligned with the left border

dasGOLE committed:
collage-designer: changed outline behavior of selected elements

dasGOLE committed:
collage-designer: cleanup of index.php

dasGOLE committed:
collage-designer: added effects to the toolbar

- added function to update state of distribution buttons

dasGOLE committed:
collage-designer: moved `lock_aspect_ratio` to the tools section

- renamed `showElmntOutline` to `showElmntOutlineBtn`

dasGOLE committed:
collage-designer: rewritten and implemented `general-settings`

allow design specific
- background img
- background color
- Frame (single over whole collage) with its own checkbox

Todo: change Demensions

dasGOLE committed:
collage-designer: fix font loading error

dasGOLE committed:
collage-designer: Added functions to adjust the dimensions of the collage.

dasGOLE committed:
collage-designer: Renamed files so that they match the js files

dasGOLE committed:
collage-designer: improved robustness of `drawCanvas` -> `draw global frame`

dasGOLE committed:
collage-designer: massiv refactoring

added:
- `collage-designer-history.js`
- `collage-designer-mouseEvents.js`
- `collage-designer-utils.js`

cleaned up:
- `collage-designer.js`
- `collage-designer-tolls.js`

dasGOLE committed:
collage-designer: removed old collage:generator refrences in new designer

dasGOLE committed:
collage-designer: fix cs ci

dasGOLE commented:
### **V1 of the designer reached.**

I've been working on further features for the past few weeks, but since I haven't had time to reintegrate the back button or deal with the latest commits, it took a while to push the latest status. @reloxx13 I had to remove your commits regarding the removal of the back button and opening the designer in a new tab. Therefore, I adjusted the history and removed them.

However, I believe all fundamental functions required for the designer are now included. There are still a few design decisions on how certain things should be handled in the future:

Open "questions":
- How to handle orientation?
- Polaroid effect will not be implemented, nor will image filters... not needed, right? -> Maybe a note that these are only visible in testing?
- Drawing a cut line (4x...designes) -> should it be included in the JSON?
- Background image now adjustable per design, possibly with a fallback to config if none is set in the design settings (how to handle this in the frontend?)
- Frame now per design (regarding the single frame, should it be possible to select a different frame for each placeholder? Currently, the global one is selected, no individual selection possible)
- CSS for the tools buttons (top) could be improved (hover, pressed states)

Currently, the designer requires the following JSON structure, for example:
```
{
"name": "1+2 Layout",
"author": "PhotoboothProject",
"aspect_ratio": "3:2",
"width": "1800",
"height": "1200",
"background": null,

"background_color": "#FFFFFF",

"elements": [
{
"type": "image",
"id": "img-0",
"x": "0",
"y": "y0.055",
"width": "x0.55546",
"height": "y0.55546",
"rotation": "10",
"show_frame": true,

"src": "resources/img/demo/seal-station-norddeich-02.jpg"

},
{
"type": "image",
"id": "img-1",
"x": "x0.555",
"y": "y0.055",
"width": "x0.40812",
"height": "y0.40812",
"rotation": "0",
"show_frame": false,
"src": null
},
{
"type": "image",
"id": "img-2",
"x": "x0.555",
"y": "y0.5368",
"width": "x0.40812",
"height": "y0.40812",
"rotation": "0",
"show_frame": false,
"src": null
},
{
"type": "text",
"id": "txt-0",
"x": "x0.5",
"y": "y0.5",
"width": "x0.3",
"height": "y*0.1",
"rotation": "30",
"content": "I'm a Test",
"font_family": "resources/fonts/GreatVibes-Regular.ttf",
"font_color": "#000000",
"font_size": 20,
"text_horizontal_align": "center",
"text_vertical_align": "center",
"font_bold": false,
"font_italic": false,
"font_underline": false
}
]
}
```

The new JSON structure could also store the background image, the single frame, and it might be possible to set a custom/different frame for each placeholder (this would then need to be supplemented in the JSON structure and the designer).
Therefore, it could look like this:
```
New JSON:
{
"width": "1800",
"height": "1200",
"background": "/resources/img/background.png",
"background_color": "#FFFFFF",
"frame": "resources/img/frames/frame.png",
"show_frame": true,
"elements": [ // NEW: Unified array for all elements
{
"type": "image",
"id": "img-1",
"x": "0", "y": "y * 0.055", "width": "1.5 * y * 0.55546", "height": "y * 0.55546", "rotation": "10",
"apply_frame": true,
"src": "placeholder.png" // Could be a generic placeholder or empty for backend to fill
},
{
"type": "text",
"id": "txt-line1",
"x": "200", "y": "220", "width": "300", "height": "50", "rotation": "10",
"content": "This is",
"font_family": "/resources/fonts/GreatVibes-Regular.ttf",
"font_color": "#420C09",
"font_size": "50",
"text_align": "center"
}
// ... more elements ...
]
}
```

Currently, the designer is not functional due to conflicts with CollageLayoutEnum. Additionally, a rebase onto the current `dev` branch would be necessary, but I would wait until "Allow custom collage layouts from private folder #1428" is merged, as this would also resolve the errors related to CollageLayoutEnum.

I'm eager for your feedback, and if this looks good, we could now proceed to the backend and the migration to the new JSON structure. :)
